File: Networking/src/main/java/jsonprotocol/ChatClientJsonWorker.java
package jsonprotocol;

import com.google.gson.Gson;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.utils.ServerException;
import org.model.Cursa;
import org.model.Inscriere;
import org.model.Participant;
import org.model.User;
import org.service.AppException;
import org.service.IClientObserver;
import org.service.IService;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

public class ChatClientJsonWorker implements Runnable, IClientObserver {
    private IService service;
    private Socket connection;
    private static Response okResponse=JsonProtocolUtils.createOkResponse();


    private BufferedReader input;
    private PrintWriter output;
    private Gson gsonFormatter;
    private volatile boolean connected;

    private static Logger logger = LogManager.getLogger(ChatClientJsonWorker.class);

    public ChatClientJsonWorker(IService server, Socket connection) {
        this.service = server;
        this.connection = connection;
        gsonFormatter=new Gson();
        try{
            output=new PrintWriter(connection.getOutputStream());
            input=new BufferedReader(new InputStreamReader(connection.getInputStream()));
            connected=true;
        } catch (IOException e) {
            logger.error(e);
            logger.error(e.getStackTrace());
        }
    }

    public void run() {
        while(connected){
            try {
                String requestLine=input.readLine();
                Request request=gsonFormatter.fromJson(requestLine, Request.class);
                Response response= handleRequest(request);
                if (response!=null){
                    sendResponse(response);
                }
            } catch (IOException e) {
                logger.error(e);
                logger.error(e.getStackTrace());
            }
            try {
                Thread.sleep(50);
            } catch (InterruptedException e) {
                logger.error(e);
                logger.error(e.getStackTrace());
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            logger.error("Error "+e);
        }
    }

    private Response handleRequest(Request request) {
        Response response = null;
        logger.info("Start handle request " + request.getType().toString() );
        if (request.getType() == RequestType.LOGIN) {
            logger.debug("Login request ...{}" + request.getUser());
            User user = request.getUser();
            try {
                user = service.userExists(user.getUsername(), user.getPassword(), this);
                return JsonProtocolUtils.createUserResponse(user);
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.LOGOUT) {
            logger.debug("Logout request ...{}" + request.getUser());
            User user = request.getUser();
            try {
                service.logout(user);
                return okResponse;
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.ADAUGA_USER) {
            logger.debug("Adaugare user ...{}" + request.getUser());
            User user = request.getUser();
            try {
                service.insertUser(user.getUsername(), user.getPassword());
                return okResponse;
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.ADAUGA_PARTICIPANT) {
            logger.debug("adauga participant ...{}" + request.getParticipant());
            Participant p = request.getParticipant();
            try {
                service.inscrieParticipant(p.getNume(),p.getCnp(),p.getCapMotor(),p.getEchipa());
                return okResponse;
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.ADAUGA_INSCRIERE) {
            logger.debug("adauga inscriere ...{}" + request.getParticipant() + " " + request.getCursa());
            Participant p = request.getParticipant();
            Cursa c = request.getCursa();
            try {
                service.adaugaInscriere(p,c);
                return okResponse;
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_ALL_TEAMS) {
            try {
                return JsonProtocolUtils.createGetTeamResponse(service.getAllTeams());
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_CAPACITATI) {
            try {
                return JsonProtocolUtils.createGetCapacitatiResponse(service.getCapacitati());
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_ALL_CURSE) {
            try {
                return JsonProtocolUtils.createCurseResponse(service.getAllCurse());
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_CURSE_BY_PARTICIPANT) {
            try {
                return JsonProtocolUtils.createCurseResponse(service.getCurseForParticipant(request.getParticipant()));
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_ALL_PARTICIPANTI) {
            try {
                return JsonProtocolUtils.createParticipantiResponse(service.getAllParticipanti());
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_CURSE_BY_CAP) {
            try {
                return JsonProtocolUtils.createCurseResponse(service.getAllCurseByCapMotor(request.getCapMotor()));
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        if (request.getType() == RequestType.GET_PARTICIPANTI_BY_ECHIPE) {
            try {
                return JsonProtocolUtils.createParticipantiResponse(service.getAllParticipantiByEchipa(request.getEchipa()));
            } catch (Exception e) {
                connected = false;
                return JsonProtocolUtils.createErrorResponse(e.getMessage());
            }
        }
        return response;
    }

    private void sendResponse(Response response) throws IOException{
        String responseLine=gsonFormatter.toJson(response);
        logger.debug("sending response "+responseLine);
        synchronized (output) {
            output.println(responseLine);
            output.flush();
        }
    }

    @Override
    public void refresh() {
        logger.info("sending refresh response");
        Response r = JsonProtocolUtils.createUpdateResponse();
        try{
            sendResponse(r);
        } catch (Exception e) {
            logger.info("error sending refresh response");
        }
    }
}

File: RaceAppC#/networking/jsonprotocol/ChatClientJsonWorker.cs
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Text.Json;
using model;
using services;
using log4net;
using services;
using log4net.Repository.Hierarchy;
using log4net.Core;
using System.Text.Json.Serialization;

namespace networking.jsonprotocol;

public class ChatClientJsonWorker : IClientObserver
{
    private IService server;
    private TcpClient connection;

    private NetworkStream stream;
    private volatile bool connected;
    private static readonly ILog log = LogManager.GetLogger(typeof(ChatClientJsonWorker));
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        //ReferenceHandler = System.Text.Json.Serialization.ReferenceHandler.Preserve,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        Converters = { new JsonStringEnumConverter() }
    };

    private static Response okResponse = JsonProtocolUtils.CreateOkResponse();
    public ChatClientJsonWorker(IService server, TcpClient connection)
    {
        this.server = server;
        this.connection = connection;
        try
        {
            stream = connection.GetStream();
            connected = true;
        }
        catch (Exception e)
        {
            log.Error(e.StackTrace);
        }
    }

    public virtual void run()
    {
        using StreamReader reader = new StreamReader(stream, Encoding.UTF8);
        while (connected)
        {
            try
            {
                string requestJson = reader.ReadLine(); // Read JSON line-by-line
                if (string.IsNullOrEmpty(requestJson)) continue;
                log.DebugFormat("Received json request {0}", requestJson);
                Request request = JsonSerializer.Deserialize<Request>(requestJson, JsonOptions);
                log.DebugFormat("Deserializaed Request {0} ", request);
                Response response = handleRequest(request);
                if (response != null)
                {
                    sendResponse(response);
                }
            }
            catch (Exception e)
            {
                log.ErrorFormat("run eroare {0}", e.Message);
                if (e.InnerException != null)
                    log.ErrorFormat("run inner error {0}", e.InnerException.Message);
                log.Error(e.StackTrace);
            }

            try
            {
                Thread.Sleep(20);
            }
            catch (Exception e)
            {
                log.Error(e.StackTrace);
            }
        }
        try
        {
            stream.Close();
            connection.Close();
        }
        catch (Exception e)
        {
            log.Error("Error " + e);
        }
    }


    private Response handleRequest(Request request)
    {
        Response response = null;

        if (request.Type == RequestType.LOGIN)
        {
            log.Debug("Login request ...{}" + request.User);
            User user = request.User;
            try
            {
                user = server.UserExists(user.Username, user.Parola, this);
                return JsonProtocolUtils.CreateUserResponse(user);
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.LOGOUT)
        {
            log.Debug("Logout request ...{}" + request.User);
            User user = request.User;
            try
            {
                server.Logout(user);
                return JsonProtocolUtils.CreateOkResponse();
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.ADAUGA_USER)
        {
            log.Debug("Adaugare user ...{}" + request.User);
            User user = request.User;
            try
            {
                server.InsertUser(user.Username, user.Parola);
                return JsonProtocolUtils.CreateOkResponse();
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.ADAUGA_PARTICIPANT)
        {
            log.Debug("Adauga participant ...{}" + request.Participant);
            Participant participant = request.Participant;
            try
            {
                server.InscrieParticipant(participant.Nume, participant.Cnp, participant.CapMotor, participant.Echipa);
                return JsonProtocolUtils.CreateOkResponse();
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.ADAUGA_INSCRIERE)
        {
            log.Debug("Adauga inscriere ...{}" + request.Participant + " " + request.Cursa);
            Participant participant = request.Participant;
            Cursa cursa = request.Cursa;
            try
            {
                server.AdaugaInscriere(participant, cursa);
                return JsonProtocolUtils.CreateOkResponse();
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.GET_ALL_TEAMS)
        {
            try
            {
                return JsonProtocolUtils.CreateGetTeamResponse(server.GetAllTeams());
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.GET_CAPACITATI)
        {
            try
            {
                return JsonProtocolUtils.CreateGetCapacitatiResponse(server.GetCapacitati());
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.GET_ALL_CURSE)
        {
            try
            {
                return JsonProtocolUtils.CreateCurseResponse(server.GetAllCurse());
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.GET_CURSE_BY_PARTICIPANT)
        {
            try
            {
                return JsonProtocolUtils.CreateCurseResponse(server.GetCurseForParticipant(request.Participant));
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.GET_ALL_PARTICIPANTI)
        {
            try
            {
                return JsonProtocolUtils.CreateParticipantiResponse(server.GetAllParticipanti());
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.GET_CURSE_BY_CAP)
        {
            try
            {
                return JsonProtocolUtils.CreateCurseResponse(server.GetAllCurseByCapMotor(request.CapMotor));
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        if (request.Type == RequestType.GET_PARTICIPANTI_BY_ECHIPE)
        {
            try
            {
                return JsonProtocolUtils.CreateParticipantiResponse(server.GetAllParticipantiByEchipa(request.Echipa));
            }
            catch (Exception e)
            {
                return JsonProtocolUtils.CreateErrorResponse(e.Message);
            }
        }

        return response;
    }


    private void sendResponse(Response response)
    {
        //de modificat pentru json
        string jsonString = JsonSerializer.Serialize(response, JsonOptions);
        log.DebugFormat("sending response {0}", jsonString);
        lock (stream)
        {
            byte[] data = Encoding.UTF8.GetBytes(jsonString + "\n"); // Append newline 
            stream.Write(data, 0, data.Length);
            stream.Flush();
        }

    }

    public void refresh()
    {
        log.Info("sending refresh response");
        Response r = JsonProtocolUtils.CreateUpdateResponse();
        try
        {
            sendResponse(r);
        }
        catch (Exception e)
        {
            log.Error("error sending refresh response from worker");
        }
    }
}